ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -311,6 +317,14 @@ private BlockImportResult importBlockAndState(
       updateForkChoiceForImportedBlock(block, result, forkChoiceStrategy);
     } else {
       result = BlockImportResult.optimisticallySuccessful(block);
+      if (optimisticSyncing.compareAndSet(false, true)
+          && recentChainData
+              .getBestBlockRoot()
+              .map(bestRoot -> bestRoot.equals(block.getParentRoot()))
+              .orElse(false)) {
+        optimisticSyncSubscribers.deliver(
+            OptimisticSyncSubscriber::onOptimisticSyncingChanged, true);
+      }
```

Comment:
I wonder if it would be simpler to detect when the chain head is optimistic by adding that info to the `ForkChoiceState` returned by `ForkChoiceStrategy.getForkChoiceState` which is called from `ForkChoice.notifyForkChoiceUpdated`.  We know that `notifyForkChoiceUpdated` is always called when the optimistic chain head changes and it's simple for the protoarray to tell if the head block it chose is optimistically sync'd or not.  Then we'd have a single place in the code that updates `optimisticSyncing` and notifies subscribers when it changes.

Comparing the best block root to the new block parent root makes me a little nervous as `recentChainData.updateChainHead` is async so it's possible we'd miss a switch to optimistic sync with this condition and it only ever detects the first optimistically sync'd block so wouldn't recover.

Reply:
yeah I wasn't proud of the solution, I just put it there to start this discussion :)

now I implemented an inner class holding `ForkChoiceState` and the `isHeadOptimistic` boolean.
I bounced back and forth from this solution and the one adding the boolean directly in the `ForkChoiceState` class.

I ended up with this to avoid adding complexity in the ForkChoiceNotifier area, where this information is not needed, even if it was the right place TBH.

Reply:
oh sorry read this comment then forgot about it in the jcenter fun and commented about putting the boolean in `ForkChoiceState`.  I think I'd still put it in `ForkChoiceState` to be honest but keep the subscribers where they are in `ForkChoice` so `ForkChoiceNotifier` will stay unchanged.  It means an extra boolean is flowing through with that `ForkChoiceState` instance but I'm ok with that - it makes sense from a design perspective that the fork choice state includes whether the head is optimistic or not and it's only one extra boolean so probably less cost than creating the extra wrapper object.

Reply:
another option is to put it in as Optional<Boolean>, so it becomes empty as soon as it is regenerated by  ForkChoiceNotifier. But let me do the right thing first :)

Reply:
wait, the booster affected my head, i confused `ForkChoiceState` with `ForkChoiceStateData`. putting it here is a nobrainer.